### DISCUSSION TITLE
pyod for EEG outlier detection

### DISCUSSION THREAD

**Discussion by EEGSC** [yzhao062/pyod#448] [Q&A]
I have segments of EEG data as 19 epochs:
```
np.shape(epoch)
Out: (19, 2)
```
For ith epoch, epoch[i][0] is the data and epoch [i][1] is the time. So,

```
np.shape(epoch[0][0])
Out: (5, 127501)
```
is the data from five EEG channels corresponding to time array

```
np.shape(epoch[0][1])
Out: (127501,)
```

I want to run a for loop 19 times in which I perform outlier detection on each epoch. I am new to this, so I simply tried the first method recommended on the documentation:

```
clf = ECOD()
clf.fit(epoch[0])

# get outlier scores
y_train_scores = clf.decision_scores_ 
```

However, I get the error:

`Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.`

I wanted to know how can I pass my data to clf.fit(). Also, is there a particular method (other than ECOD) that one would recommend for EEG outlier detection?

**Comment by yzhao062** [comment]
I think PyOD may not be a perfect choice for time-series since it does not handle the time stamps. what about https://github.com/datamllab/tods?

**Reply by EEGSC** [reply]
I tried the following:

```
clf = ECOD()
clf.fit(np.transpose(epoch[0][0]))
#X (the input) has to be of the size (n_samples,n_features)

# get outlier scores
clf.decision_scores_  # raw outlier scores on the train data
clf.labels_  # binary labels (0: inliers, 1: outliers)
z=clf.labels_.tolist()
z.count(1) #Counting the number of 1s, ie. the number of outliers
#Percentage of outliers in the data
(z.count(1)/len(z))*100
```
The last line had the output 9.999921569242593. So I think I did get that in this particular epoch, 10% of the data are outliers. Wouldn't thinking of epochs as samples, and number of channels as features work? I think I can stack the epochs vertically and pass. Wouldn't that serve the goal?
